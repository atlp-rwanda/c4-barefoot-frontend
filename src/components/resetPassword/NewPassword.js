import React from 'react';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import Container from '@material-ui/core/Container';
import { ToastContainer, toast, Zoom, Bounce } from 'material-react-toastify';
import { useDispatch, connect, useSelector } from 'react-redux';
import { resetNewPassword } from '../../redux/actions/resetPasswordAction';
import { Field, Form, Formik } from 'formik'
import { FormGroup, Snackbar } from '@material-ui/core';
import { object, string, ref } from 'yup';
import MuiAlert from '@material-ui/lab/Alert';

const useStyles = makeStyles((theme) => ({
  paper: {
    marginTop: theme.spacing(8),
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    padding:'10px'
  },
   sendingHeader:{
            justifyContent:'center',
            alignItems:'center',
            alignContent:'center',
            marginTop:'40vh',
            color:'gray',
            textAlign:'center'
        },
  avatar: {
    margin: theme.spacing(1),

    backgroundColor: theme.palette.secondary.main,
  },
  errors:{
      color:'red'
      },

  form: {
    width: '100%', // Fix IE 11 issue.
    marginTop: theme.spacing(1),
    padding:'20px',
    height:'fit-content',
    border: '1px solid lightgray',
    borderBottom: '2px solid lightgray',
    borderRadius: '10px'
  },
  submit: {
    margin: theme.spacing(3, 0, 2),
    float:'right',
    marginRight:0
  },
}));
  
const initialValues = {
  password:'',
  confirmPassword:''
}

function NewPassword(props) {
  const classes = useStyles();

   const { history } = props
   const query = history.location.search
   const errors = props.newpassword.error
    const opened = props.newpassword.open
    const anyLoading = props.newpassword.isLoading
    const messages = props.newpassword.message

  const handleSubmition = (values)=>{
        event.preventDefault()
        props.resetNewPassword(values, query)
        values=''
    };
  function handleClose(event, reason){
        if(reason === 'clickaway'){
            return opened;
        }
    };
   if(props.newpassword.success){
     props.history.push('/login');
   }
    function Alert(props){
        return < MuiAlert elevation={6} variant="filled" {...props}/>
    };
  return (
    <>
    {errors ? (<Snackbar open={opened} autoHideDuration={5000} onclose={handleClose}>
                <Alert onclose={handleClose} severity="error">
                    Error: {errors.error? (errors.error): (<div>{errors}</div>)}
                </Alert>
            </Snackbar>): (<Snackbar open={opened} autoHideDuration={5000} onclose={handleClose}>
                <Alert onclose={handleClose} severity="success">
                    successful: {messages}üëç You can login with new Password 
                </Alert>
            </Snackbar>)}
    <Container data-test='container' component="main" maxWidth="xs">
        <ToastContainer />
      {anyLoading? ( <Typography className={classes.sendingHeader} variant="h5" > Sending ...</Typography> ): (
      <div className={classes.paper}>
        <Typography component="h1" variant="h5">
          Reset Password
        </Typography>
        <Formik
        initialValues={initialValues}
        validationSchema={
          object().shape({
            password:string().required(),
            confirmPassword:string().required().when("password", {
               is: val => (val && val.length > 0 ? true : false),
               then: string().oneOf(
               [ref("password")], "Both password need to be the same")
            })
          })
        }
        onSubmit={handleSubmition}
        >
        {({ values, handleChange, handleBlur, errors, touched }) => (
            <Form data-test="form-test" className={classes.form} noValidate>
                <FormGroup>
                  <Field
                    as={TextField}
                    variant="outlined"
                    margin="normal"
                    value={values.password}
                    onChange={handleChange}
                    onBlur={handleBlur}
                    required
                    fullWidth
                    type="password"
                    id="password"
                    label="New password"
                    name="password"
                    autoComplete="current-password"
                    autoFocus
                  />
                  {touched.password && errors.password ? (<span className={classes.errors} >{errors.password }</span>): null}
                </FormGroup>
                <FormGroup>
                  <Field
                    as={TextField}
                    variant="outlined"
                    margin="normal"
                    value={values.confirmPassword}
                    onChange={handleChange}
                    onBlur={handleBlur}
                    required
                    fullWidth
                    name="confirmPassword"
                    label="Confirm Password"
                    type="password"
                    id="confirmPassword"
                    autoComplete="current-password"
                  />
                  {touched.confirmPassword && errors.confirmPassword ? (<span className={classes.errors} >{errors.confirmPassword }</span>): null}
                </FormGroup>
              <Button
                type="submit"
                fullWidth
                variant="contained"
                color="primary"
                className={classes.submit}
                // onClick={notify}
              >
                RESET
              </Button>
            </Form>
        )}
        </Formik>
      </div>
      )}
    </Container>
    </>
  );
};

const mapStateToProps = state =>({
    newpassword: state.newPassword
});
export { NewPassword }
export default connect(mapStateToProps, {resetNewPassword})(NewPassword)