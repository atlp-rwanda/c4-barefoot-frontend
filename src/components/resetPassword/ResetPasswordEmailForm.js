import React, { useState, useEffect } from 'react'
import { useSelector, connect, useDispatch } from 'react-redux'
import { FormGroup, Grid, makeStyles, Paper, TextField, Typography, Button, CircularProgress, Snackbar, Slide, Container  } from '@material-ui/core'
import { Field, Form, Formik } from 'formik'
import { sendEmail, closeSnackbar } from "../../redux/actions/resetPasswordAction";
import { object, string } from 'yup'
import { ToastContainer, toast, Zoom, Bounce } from 'material-react-toastify';
import MuiAlert from '@material-ui/lab/Alert';

const useStyle = makeStyles( (theme) => ({
    typographyColor: {
            color:'#257AAA',
            margin:'30px auto',
            align:"center" 
        },
        loadingHeader:{
            justifyContent:'center',
            alignItems:'center',
            alignContent:'center',
            marginTop:'40vh',
            color:'gray',
            textAlign:'center'
        },
        message:{
            color:'green',
            textAlign:'center',
            padding:'10px',
            borderRadius:'8px'
        },
        rootGrid: {
            justifyContent:'center',
            height:'100vh'
        },
        errors:{
            color:'red'
        },
        buttonStyle: {
            margin: '20px auto',
            background:'#257AAA',
            width:"20vw",
            '@media(max-width:1024px)':{
                width:'100%'
            },
            '@media(max-width: 740px)' :{
                minHeight: '100%'
            },
            float:'right',
            marginRight:0,
            borderRadius:0
        },
        emailForm: {
            border: 'none'
        },
        paperStyle: {
            padding:' 5px 20px',
            maxHeight:'fit-content',
            margin:'70px auto',
            '@media(max-width: 740px)' :{
                minHeight: '450px'
            },
            '@media(max-width:1024px)':{
                marginTop:'90px'
            }
        }

}));

const initialValues = {
    email:''
}

function ResetPasswordEmailForm(props){
    // const dispatch = useDispatch()

    const [Email, setEmail] = useState({email:''})
    const [loading, SetLoading] = useState(false)
    let { email } = Email;
    const errors = props.sendemail.error
    const opened = props.sendemail.open
    const anyLoading = props.sendemail.isLoading
    const messages = props.sendemail.message
    const myerror = JSON.stringify(errors)
    const classes= useStyle();
    const notify = (errors) => {
            toast.success(<p>Your email sent successful!</p>, {position:toast.POSITION.TOP_CENTER})
        };
        
    // let { isLoading } = loading;


    const emailHandle = (e) => {
        setEmail({...Email, [e.target.name]:e.target.value});
        setLoading({...loading, isLoading:true});
    };


    useEffect(() => {
        const loadingTime = setTimeout(() => {
            SetLoading(true);
        }, 2000)
        return () => cleanTimeout(loadingTime)
    }, []);

    const handleSubmition = (values)=>{
        props.sendEmail(values)
        values=''
    };
    function handleClose(){
        props.closeSnackbar()
        //  dispatch(closeSnackbar())
    };

    function Alert(props){
        return < MuiAlert elevation={6} variant="filled" {...props} itemID='alert' />
    };
        return (
            <>
            {errors ? (<Snackbar open={opened} autoHideDuration={5000} onClose={handleClose}>
                <Alert onClose={handleClose} severity="error">
                    Error: {errors}
                </Alert>
            </Snackbar>): (<Snackbar open={opened} autoHideDuration={5000} onClose={handleClose}>
                <Alert onClose={handleClose} severity="success">
                    successful: {messages}
                </Alert>
            </Snackbar>)}
                
            <Grid container data-test="main-grid" className={classes.rootGrid} component='main'>
                {/* <ToastContainer /> */}
                
            {anyLoading? ( <Typography className={classes.loadingHeader} variant="h5" > Loading ...</Typography> ):
            (
                <Grid item xs={10} sm={8} md={6}>
                <Paper item className={classes.paperStyle} >
                    <Typography data-test="typography-test" className={classes.typographyColor} variant="h5">Forgot your password don't worry ?</Typography>
                    <Typography data-test="typography-test" variant="h6" className={classes.typographyColor}>Enter email below and send you a link to reset your password.</Typography>
                    {/* <Typography className={classes.message}>{messages}</Typography> */}
                    <Formik 
                    initialValues={initialValues}
                    validationSchema={
                        object().shape({
                            email:string().required("Email is required!").email()
                        })
                    }
                    onSubmit={handleSubmition} 
                    >

                    {({ values, handleChange, handleBlur, errors, touched }) => (
                        <Form className={classes.emailForm} 
                        noValidate
                        data-test="form-test"
                        >
                        {/* <pre>{JSON.stringify(errors, null, 4)}</pre>
                        <pre>{JSON.stringify(values, null, 4)}</pre> */}
                        <FormGroup>
                             <Field 
                             name='email' 
                             as={TextField} 
                             label='Email'
                             type='email'
                             id="standard-basic" 
                             value={values.email} 
                             name='email'
                             onChange={handleChange}
                             onBlur={handleBlur}
                             placeholder="Enter Your email" 
                             fullWidth 
                             required
                             errors={errors.email}
                             touched={touched.email}
                             autoFocus 
                             data-test="input-test"
                             />
                            {touched.email && errors.email ? (<span className={classes.errors} >{errors.email }</span>) : null } 
                        <Button 
                            variant="contained"
                            type="submit"
                            maxWidth="sm"
                            color="primary"
                            className={classes.buttonStyle}
                            >Send</Button>
                            {/* <br /> */}
                        </FormGroup>
                        </Form>
                    )}
                    </Formik>
                </Paper>
                </Grid>
            )
            }
            </Grid>
           </> 
        )
    };
    const mapStateToProps = state =>({
    sendemail: state.sendEmail
});

export default connect(mapStateToProps, {sendEmail, closeSnackbar})(ResetPasswordEmailForm)
export { ResetPasswordEmailForm}