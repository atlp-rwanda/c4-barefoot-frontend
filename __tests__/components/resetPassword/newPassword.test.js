import React from 'react';
import Enzyme, { shallow } from "enzyme";
import toJson from 'enzyme-to-json';
import EnzymeAdapter from 'enzyme-adapter-react-16';
import { Field, Form, Formik } from 'formik'
import { FormGroup, Grid, makeStyles, Paper, TextField, Typography, Button, CircularProgress, Snackbar, Slide, Container, Input  } from '@material-ui/core'
import { NewPassword } from '../../../src/components/resetPassword/NewPassword';

Enzyme.configure({ adapter: new EnzymeAdapter()});
describe('reset new password', ()=>{
    //  let wrapper;
         const props = {
             resetNewPassword: jest.fn(),
             history : {
            replace: jest.fn(),
            length: 0,
            location: { 
                pathname: '',
                search: '/user/reset_password',
                state: '',
                hash: ''
            }},
             newpassword: {
                isLoading:false,
                    open:false,
                    error: '',
                    message:''
            }
         }
     it('it match snapshoot', () => {
         const props = {
             resetNewPassword: jest.fn(),
             history : {
            replace: jest.fn(),
            length: 0,
            location: { 
                pathname: '',
                search: '/user/reset_password',
                state: '',
                hash: ''
            }},
             newpassword: {
                isLoading:false,
                    open:false,
                    error: '',
                    message:''
            }
         }
        //  const history = createMemoryHistory('/user/reset_password')
          const wrapper = shallow(<NewPassword  {...props} />)
         const component = wrapper.find("[data-test='container']");
         expect(toJson(wrapper)).toMatchSnapshot();
     });
   it('should render container component', ()=>{
       const props = {
             resetNewPassword: jest.fn(),
             history : {
            replace: jest.fn(),
            length: 0,
            location: { 
                pathname: '',
                search: '/user/reset_password',
                state: '',
                hash: ''
            }},
             newpassword: {
                isLoading:false,
                    open:false,
                    error: '',
                    message:''
            }
         }
       const wrapper = shallow(<NewPassword  {...props} />)
        const component = wrapper.find("[data-test='container']");
        expect(component.length).toBe(1)
   });
   it('should render container component', ()=>{
       const props = {
             resetNewPassword: jest.fn(),
             history : {
            replace: jest.fn(),
            length: 0,
            location: { 
                pathname: '',
                search: '/user/reset_password',
                state: '',
                hash: ''
            }},
             newpassword: {
                isLoading:false,
                    open:false,
                    error: '',
                    message:''
            }
         }
       const wrapper = shallow(<NewPassword  {...props} />)
        const component = wrapper.find(Typography);
        expect(component.length).toBe(1)
        expect(component.text()).toBe('Reset Password')
   });
   it('should render Formik component', ()=>{
       const props = {
             resetNewPassword: jest.fn(),
             history : {
            replace: jest.fn(),
            length: 0,
            location: { 
                pathname: '',
                search: '/user/reset_password',
                state: '',
                hash: ''
            }},
             newpassword: {
                isLoading:false,
                    open:false,
                    error: '',
                    message:''
            }
         }
       const wrapper = shallow(<NewPassword  {...props} />)
        const component = wrapper.find(Formik);
        expect(component.length).toBe(1)
   });

    
 })