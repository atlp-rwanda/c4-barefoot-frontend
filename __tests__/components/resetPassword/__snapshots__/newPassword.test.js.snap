// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reset new password it match snapshoot 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Snackbar))
    autoHideDuration={5000}
    onclose={[Function]}
    open={false}
  >
    <Alert
      onclose={[Function]}
      severity="success"
    >
      successful: 
      üëç You can login with new Password
    </Alert>
  </WithStyles(ForwardRef(Snackbar))>
  <WithStyles(ForwardRef(Container))
    component="main"
    data-test="container"
    maxWidth="xs"
  >
    <ToastContainer
      autoClose={3000}
      closeButton={[Function]}
      closeOnClick={true}
      draggable={true}
      draggablePercent={40}
      hideProgressBar={true}
      newestOnTop={false}
      pauseOnFocusLoss={true}
      pauseOnHover={true}
      position="bottom-left"
      role="alert"
      rtl={false}
      transition={[Function]}
    />
    <div
      className="makeStyles-paper-1"
    >
      <WithStyles(ForwardRef(Typography))
        component="h1"
        variant="h5"
      >
        Reset Password
      </WithStyles(ForwardRef(Typography))>
      <Formik
        initialValues={
          Object {
            "confirmPassword": "",
            "password": "",
          }
        }
        onSubmit={[Function]}
        validationSchema={
          ObjectSchema {
            "_blacklist": ReferenceSet {
              "list": Set {},
              "refs": Map {},
            },
            "_blacklistError": undefined,
            "_excludedEdges": Array [],
            "_nodes": Array [
              "password",
              "confirmPassword",
            ],
            "_sortErrors": [Function],
            "_typeError": [Function],
            "_whitelist": ReferenceSet {
              "list": Set {},
              "refs": Map {},
            },
            "_whitelistError": undefined,
            "conditions": Array [],
            "deps": Array [],
            "exclusiveTests": Object {},
            "fields": Object {
              "confirmPassword": StringSchema {
                "_blacklist": ReferenceSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_blacklistError": undefined,
                "_typeError": [Function],
                "_whitelist": ReferenceSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_whitelistError": undefined,
                "conditions": Array [
                  Condition {
                    "fn": [Function],
                    "refs": Array [
                      Reference {
                        "getter": [Function],
                        "isContext": false,
                        "isSibling": true,
                        "isValue": false,
                        "key": "password",
                        "map": undefined,
                        "path": "password",
                      },
                    ],
                  },
                ],
                "deps": Array [
                  "password",
                ],
                "exclusiveTests": Object {
                  "required": true,
                },
                "spec": Object {
                  "abortEarly": true,
                  "label": undefined,
                  "meta": undefined,
                  "nullable": false,
                  "presence": "required",
                  "recursive": true,
                  "strict": false,
                  "strip": false,
                },
                "tests": Array [
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
                "type": "string",
              },
              "password": StringSchema {
                "_blacklist": ReferenceSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_blacklistError": undefined,
                "_mutate": undefined,
                "_typeError": [Function],
                "_whitelist": ReferenceSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_whitelistError": undefined,
                "conditions": Array [],
                "deps": Array [],
                "exclusiveTests": Object {
                  "required": true,
                },
                "spec": Object {
                  "abortEarly": true,
                  "label": undefined,
                  "meta": undefined,
                  "nullable": false,
                  "presence": "required",
                  "recursive": true,
                  "strict": false,
                  "strip": false,
                },
                "tests": Array [
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
                "type": "string",
              },
            },
            "spec": Object {
              "abortEarly": true,
              "label": undefined,
              "meta": undefined,
              "nullable": false,
              "presence": "optional",
              "recursive": true,
              "strict": false,
              "strip": false,
            },
            "tests": Array [],
            "transforms": Array [
              [Function],
            ],
            "type": "object",
          }
        }
      >
        <Component />
      </Formik>
    </div>
  </WithStyles(ForwardRef(Container))>
</Fragment>
`;
